chore: add Gen2 specific doc for using SSR adapter with Nuxt 3 (#7575)
aws-amplify/docs: src/pages/[platform]/build-a-backend/server-side-rendering/nuxt/index.mdx

                      libraryOptions,
                      (contextSpec) => gqlServerClient.models.Todo.update(contextSpec, ...input)
                    )
                  }
                }
              }
            }
          }
        }
      }
    };
  }
});
```

#### Usage example

Using the Storage `list` API in `pages/storage-list.vue`:

```ts title="pages/storage-list.vue"

// `useAsyncData` and `useNuxtApp` are Nuxt composables
// `$Amplify` is generated by Nuxt according to the `provide` key in the plugins
// we've added above
<script setup lang="ts">
const { data, error } = useAsyncData(async () => {
  const listResult = await useNuxtApp().$Amplify.Storage.list({
    path: 'album'
  });
  return listResult.items;
});
</script>

<template>
  <h3>Files under path: album</h3>
  <pre>{{ data }}</pre>
</template>
```

Using the GraphQL API in `pages/todos-list.vue`:

```ts title="pages/todos-list.vue"
<script setup lang="ts">
const { data, error } = useAsyncData(async () => {
  const { data } = await useNuxtApp().$Amplify.GraphQL.client.models.Todo.list();
  return data;
});
</script>

<template>
  <h3>Todos</h3>
  <pre>{{ data }}</pre>
</template>
```

> The above two pages can be rendered on both the client and server sides by default. `useNuxtApp().$Amplify` will pick up the correct implementation of `01.amplifyApis.client.ts` and `01.amplifyApis.server.ts` to use, depending on the runtime.

Reviewer: nit
```suggestion
The above two pages can be rendered on both the client and server by default. `useNuxtApp().$Amplify` will pick up the correct implementation of `01.amplifyApis.client.ts` and `01.amplifyApis.server.ts` to use, depending on the runtime.
```

<Callout warning>

Only a subset of Amplify APIs are usable on the server side, and as the libraries evolve, `amplify-apis.client` and `amplify-apis.server` may diverge further. You can guard your API calls to ensure an API is available in the context where you use it (e.g., you can use `if (process.client)` to ensure that a client-only API isn't executed on the server).

</Callout>

## Set up Auth middleware to protect your routes

> The auth middleware will use the plugin set up in the previous step as a dependency; therefore you can add the auth middleware via another plugin that will be loaded after the previous one.

Reviewer: ```suggestion
The auth middleware will use the plugin configured in the previous step as a dependency; therefore you can add the auth middleware via another plugin that will be loaded after the previous one.
```

import { getCustomStaticPath } from '@/utils/getCustomStaticPath';
import { getChildPageNodes } from '@/utils/getChildPageNodes';

export const meta = {
  title: 'Use Amplify categories APIs from Nuxt 3',
  description: 'Use Amplify categories APIs from Nuxt 3',
  platforms: [
    'javascript',
    'vue'
  ],
};

export const getStaticPaths = async () => {
  return getCustomStaticPath(meta.platforms);
};

export function getStaticProps(context) {
  const childPageNodes = getChildPageNodes(meta.route);
  return {
    props: {
      platform: context.params.platform,
      meta,
      childPageNodes
    }
  };
}

> If you have not already done so, please read the introduction documentation, [Use Amplify Categories APIs in Server Side Rendering](/gen1/[platform]/build-a-backend/server-side-rendering/), to learn about how to use Amplify categories APIs in server-side rendering.

Reviewer 2: ```suggestion
If you have not already done so, please read the introduction documentation, [Use Amplify Categories APIs in Server-Side Rendering](/gen1/[platform]/build-a-backend/server-side-rendering/), to learn about how to use Amplify categories' APIs in server-side rendering.
```

> This documentation provides a getting-started guide to using the generic `runWithAmplifyServerContext` adapter (exported from `aws-amplify/adapter-core`) to enable Amplify Gen2 in a Nuxt 3 project. The examples in this documentation may not present best practices for your Nuxt project. You are welcome to provide suggestions and contributions to improve this documentation or to create a Nuxt adapter package for Amplify and let others use it.

Reviewer 2: ```suggestion
This documentation provides a getting-started guide to using the generic `runWithAmplifyServerContext` adapter (exported from `aws-amplify/adapter-core`) to enable Amplify in a Nuxt 3 project. The examples in this documentation may not present best practices for your Nuxt project. You are welcome to provide suggestions and contributions to improve this documentation or to create a Nuxt adapter package for Amplify and let others use it.
```

<Callout>

> **NOTE:** This guide assumes that you have a deep knowledge of Nuxt 3.

Reviewer 2: ```suggestion
**Note:** This guide assumes that you have a deep knowledge of Nuxt 3.
```

</Callout>

> ## Start Using Amplify Gen2 in Your Nuxt 3 project

Reviewer 2: can we remove all "gen2" references and remove the Title Case in favor of Sentence case for headings?

> You can install Amplify Gen2 following this guide, [Manual installation](/[platform]/start/manual-installation/).

Reviewer 2: ```suggestion
You can install relevant Amplify libraries by following the [manual installation](/[platform]/start/manual-installation/) guide.
```

## Set Up the AmplifyAPIs Plugin

Nuxt 3 offers universal rendering by default, where your data fetching logic may be executed on both the client and server sides. Amplify offers APIs that are capable of running within a server context to support use cases such as server-side rendering (SSR) and static site generation (SSG), though Amplify's client-side APIs and server-side APIs of Amplify are slightly different. You can set up an `AmplifyAPIs` plugin to make your data fetching logic run smoothly across the client and server.

1. If you haven’t already done so, create a `plugins` directory under the root of your Nuxt project
2. Create two files `01.amplify-apis.client.ts` and `01.amplify-apis.server.ts` under the `plugins` directory

<Callout>

> **NOTE:** the leading number in the files name indicate the plugin loading order, details see https://nuxt.com/docs/guide/directory-structure/plugins#registration-order. The `.client` and `.server` indicate the runtime that the logic contained in the file will run on, client or server. For details see: https://nuxt.com/docs/guide/directory-structure/plugins

Reviewer 2: ```suggestion
**Note:** the leading number in the files name indicate the plugin loading order, details see https://nuxt.com/docs/guide/directory-structure/plugins#registration-order. The `.client` and `.server` indicate the runtime that the logic contained in the file will run on, client or server. For details see: https://nuxt.com/docs/guide/directory-structure/plugins
```

to align with other "note" callouts

Make sure you call `Amplify.configure` as early as possible in your application’s life-cycle. A missing configuration or `NoCredentials` error is thrown if `Amplify.configure` has not been called before other Amplify JavaScript APIs. Review the [Library Not Configured Troubleshooting guide](/gen1/[platform]/build-a-backend/troubleshooting/library-not-configured/) for possible causes of this issue.

</Callout>

### Implement `01.amplify-apis.server.ts`

Example implementation:

```ts title="plugins/01.amplify-apis.server.ts"
import type { CookieRef } from 'nuxt/app';
import {
  createKeyValueStorageFromCookieStorageAdapter,
  createUserPoolsTokenProvider,
  createAWSCredentialsAndIdentityIdProvider,
  runWithAmplifyServerContext
} from 'aws-amplify/adapter-core';
import { parseAmplifyConfig } from 'aws-amplify/utils';
import {
  fetchAuthSession,
  fetchUserAttributes,
  getCurrentUser
} from 'aws-amplify/auth/server';
import { list } from 'aws-amplify/storage/server';
import { generateClient } from 'aws-amplify/api/server';
import type { ListPaginateWithPathInput } from 'aws-amplify/storage';
import type {
  LibraryOptions,
  FetchAuthSessionOptions
} from '@aws-amplify/core';

> import type { Schema } from '~/amplify/data/resource';

Reviewer 2: nit, can we group all of the type imports and lift to the top of the import block?

> import config from '../amplify_outputs.json';
> 
> // parse the content of `amplify_outputs.json` into the shape of ResourceConfig
> const amplifyConfig = parseAmplifyConfig(config);

Reviewer 2: ```suggestion
import outputs from '../amplify_outputs.json';

// parse the content of `amplify_outputs.json` into the shape of ResourceConfig
const amplifyConfig = parseAmplifyConfig(outputs);
```

aligns with other examples

// create the Amplify used token cookies names array
const userPoolClientId = amplifyConfig.Auth!.Cognito.userPoolClientId;
const lastAuthUserCookieName = `CognitoIdentityServiceProvider.${userPoolClientId}.LastAuthUser`;

// create a GraphQL client that can be used in a server context
const gqlServerClient = generateClient<Schema>({ config: amplifyConfig });

// Extra the model operation function types for creating wrapper function later

> type RemoveFirstParam<Params extends any[]> = Params extends [infer _, ...infer Rest] ? Rest : never; 

Reviewer 2: can this be removed now in favor of `Schema["SomeModel"]["type"]`?

Author: The model operation APIs have a second parameter as well as `options`, `Schema["SomeModel"]["type"]` is insufficient for this use case.

Reviewer: For future ref: @svidgen is working on a refactor that will expose additional CRUDL input types directly on the `Schema` type. E.g. you'll be able to call `Schema['ModelName']['createType']`. Full list [here](https://github.com/aws-amplify/amplify-api-next/pull/206/files#diff-fc8718b93a5b01e4886e15d9a14390a2fe8025a7b712bd793a23afc4b11ca151R30-R36). Let us know if this will solve your intended use case.

                  },
                  create(...input: TodoCreateInput) {
                    return runWithAmplifyServerContext(
                      amplifyConfig,
                      libraryOptions,
                      (contextSpec) => gqlServerClient.models.Todo.create(contextSpec, ...input)
                    )
                  },
                  update(...input: TodoUpdateInput) {
                    return runWithAmplifyServerContext(
                      amplifyConfig,
                      libraryOptions,
                      (contextSpec) => gqlServerClient.models.Todo.update(contextSpec, ...input)
                    )
                  }
                }
              }
            }
          }
        }
      }
    };
  }
});
```

#### Usage Example

Using the Storage `list` API in `pages/storage-list.vue`:

```ts title="pages/storage-list.vue"

// `useAsyncData` and `useNuxtApp` are Nuxt composables
// `$Amplify` is generated by Nuxt according to the `provide` key in the plugins
// we've added above
<script setup lang="ts">
const { data, error } = useAsyncData(async () => {
  const listResult = await useNuxtApp().$Amplify.Storage.list({

>     path: 'public/'

Reviewer 3: nit: can we make this generic and not ref to accessLevel?

  });
  return listResult.items;
});
</script>

<template>

>   <h3>Files with access level: guest</h3>

Reviewer 3: Same here: Can we remove reference to accesslevel?

> 
> // `useAsyncData` and `useNuxtApp` are Nuxt composables
> // `$Amplify` is generated by Nuxt according to the `provide` key in the plugins
> // we've added above
> <script setup lang="ts">
> const { data, error } = useAsyncData(async () => {
>   const listResult = await useNuxtApp().$Amplify.Storage.list({
>     path: 'album'
>   });
>   return listResult.items;
> });
> </script>
> 
> <template>
>   <h3>Files with under path: album</h3>
>   <pre>{{ data }}</pre>
> </template>
> 

Reviewer 2: ```suggestion
// `useAsyncData` and `useNuxtApp` are Nuxt composables
// `$Amplify` is generated by Nuxt according to the `provide` key in the plugins
// we've added above
<script setup lang="ts">
const { data, error } = useAsyncData(async () => {
  const listResult = await useNuxtApp().$Amplify.Storage.list({
    path: 'album'
  });
  return listResult.items;
});
</script>

<template>
  <h3>Files with under path: album</h3>
  <pre>{{ data }}</pre>
</template>
```

nit for formatting/styling

```

Using the GraphQL API in `pages/todos-list.vue`:

```ts title="pages/todos-list.vue"
<script setup lang="ts">
const { data, error } = useAsyncData(async () => {
  const { data } = await useNuxtApp().$Amplify.GraphQL.client.models.Todo.list();
  return data;
});
</script>

<template>
  <h3>Todos</h3>
  <pre>{{ data }}</pre>
</template>
```

The above two pages can be rendered on both the client and server sides by default. `useNuxtApp().$Amplify` will pick up the correct implementation of `01.amplify-apis.client.ts` and `01.amplify-apis.server.ts` to use, depending on the runtime.

<Callout warning>

> Only a subset of Amplify APIs are usable on the server side, and as the libraries evolve, `amplify-apis.client` and `amplify-apis.server` may diverge further. You can guard your API calls to ensure an API is available in the context where you use it. E.g., you can use `if (process.client)` to ensure that a client-only API isn't executed on the server.

Reviewer 2: ```suggestion
Only a subset of Amplify APIs are usable on the server side, and as the libraries evolve, `amplify-apis.client` and `amplify-apis.server` may diverge further. You can guard your API calls to ensure an API is available in the context where you use it (e.g., you can use `if (process.client)` to ensure that a client-only API isn't executed on the server).
```

</Callout>

## Set Up Auth Middleware to Protect Your Routes

The auth middleware will use the plugin set up in the previous step as a dependency; therefore you can add the auth middleware via another plugin that will be loaded after the previous one.

1. Create a `02.auth-redirect.ts` file under plugins directory

<Callout>

> **NOTE:** This file will run on both client and server, details see: https://nuxt.com/docs/guide/directory-structure/middleware#when-middleware-runs. The `02` name prefix ensures this plugin loads after the previous so `useNuxtApp().$Amplify` becomes available.

Reviewer 2: ```suggestion
**Note:** This file will run on both client and server, details see: https://nuxt.com/docs/guide/directory-structure/middleware#when-middleware-runs. The `02` name prefix ensures this plugin loads after the previous so `useNuxtApp().$Amplify` becomes available.
```

</Callout>

### Implement `02.auth-redirect.ts`

Example implementation:

```ts title="plugins/02.auth-redirect.ts"
import { Amplify } from 'aws-amplify';

> import config from '~/amplify_outputs.json';

Reviewer 2: `config` -> `outputs`

</Callout>

## Set Up Amplify for API Route Use Cases

Following the specification of Nuxt, your API route handlers will live under `~/server`, which is a separate environment from other parts of your Nuxt app; hence, the plugins created in the previous sections are not usable here, and extra work is required.

### Set Up Amplify Server Context Utility

1. If you haven’t already done so, create a `utils` directory under the server directory of your Nuxt project
2. Create an `amplifyUtils.ts` file under the `utils` directory

In this file, you will create a helper function to call Amplify APIs that are capable of running on the server side with context isolation.

Example implementation:

```ts title="utils/amplifyUtils.ts"
import type { H3Event, EventHandlerRequest } from 'h3';
import {
  createKeyValueStorageFromCookieStorageAdapter,
  createUserPoolsTokenProvider,
  createAWSCredentialsAndIdentityIdProvider,
  runWithAmplifyServerContext,
  AmplifyServer,
  CookieStorage
} from 'aws-amplify/adapter-core';
import { parseAmplifyConfig } from 'aws-amplify/utils';

> import type { LibraryOptions } from '@aws-amplify/core';

Reviewer 2: same nit for grouping type imports at the top of the file. Additionally, does this break when using npm's `install-strategy=linked`, yarn, or pnpm? pnpm will not resolve imports from transient deps

Author: Tested with `install-strategy=linked` Nuxt itself started failing while running `postinstall` script. Consider the following statement printed by npm

```
npm WARN reify The "linked" install strategy is EXPERIMENTAL and may contain bugs.
```

I think we can use this import path as is for now.

Reviewer 2: is this type export available from the top-level package? pnpm is unable to resolve it
![CleanShot 2024-05-16 at 13 02 33@2x](https://github.com/aws-amplify/docs/assets/5033303/1e0d5135-dc09-4e99-9cec-b5f5fb7c99cc)

> import config from '~/amplify_outputs.json';

Reviewer 2: config -> outputs

    operation
  );
};
```

You can then use `runAmplifyApi` function to call Amplify APIs in an isolated server context.

#### Usage Example

Take implementing an API route `GET /api/current-user` , in `server/api/current-user.ts`:

```ts title="server/api/current-user.ts"
import { getCurrentUser } from 'aws-amplify/auth/server';
import { runAmplifyApi } from '~/server/utils/amplifyUtils';

export default defineEventHandler(async (event) => {
  const user = await runAmplifyApi(event, (contextSpec) =>
    getCurrentUser(contextSpec)
  );

  return user;
});
```

Then you can fetch data from this route, for example, `fetch('http://localhost:3000/api/current-user')`.

## Set Up Server Middleware to Protect Your API Routes

Similar to API routes, the previously added auth middleware are not usable under `/server`, hence extra work is required to set up a auth middleware to protect your routes.

1. If you haven’t already done so, create a `middleware` directory under the `server` directory of your Nuxt project
2. Create an `amplifyAuthMiddleware.ts` file under the `middleware` directory

This middleware will be executed before a request reach your API route.

Example implementation:

> ```ts title="middleware/amplifyAuthMiddleware.ts"

Reviewer 2: does Nuxt/Vue have a pattern for filenames? filenames earlier in the doc are kebab-case where this is camelCase

Author: I don't think so, the snake cases above followed Nuxt documentations, tested using camel case doesn't break them. I can update and unify.

<script setup lang="ts">
const { data, error } = useAsyncData(async () => {
  const listResult = await useNuxtApp().$Amplify.Storage.list({
    path: 'album'
  });
  return listResult.items;
});
</script>

<template>

>   <h3>Files with under path: album</h3>

Reviewer 3: ```suggestion
  <h3>Files under path: album</h3>
```
